core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/log.go

```diff
@@ -0,0 +1,69 @@
+package logprovider
+
+import (
+	"encoding/hex"
+
+	ocr2keepers "github.com/smartcontractkit/chainlink-common/pkg/types/automation"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+)
+
+// LogSorter sorts the logs based on block number, tx hash and log index.
+// returns true if b should come before a.
+func LogSorter(a, b logpoller.Log) bool {
+	return LogComparator(a, b) > 0
+}
+
+// LogComparator compares the logs based on block number, log index.
+// tx hash is also checked in case the log index is not unique within a block.
+//
+// Returns:
+//
+//	-1 if a <  b
+//	 0 if a == b
+//	+1 if a >  b
+func LogComparator(a, b logpoller.Log) int {
+	blockDiff := int(a.BlockNumber - b.BlockNumber)
+	if blockDiff != 0 {
+		return normalizeCompareResult(blockDiff)
+	}
+	logIndexDiff := int(a.LogIndex - b.LogIndex)
+	if logIndexDiff != 0 {
+		return normalizeCompareResult(logIndexDiff)
+	}
+	return a.TxHash.Big().Cmp(b.TxHash.Big())
```

Comment:
technically this can also be the same, since unique identifier of log (logID) is index, txHash, blockHash.

This can be an edge case during reorgs, will think more about implications

Reply:
Not sure that reorgs should be managed within this function as it complicates things, the function will be stateful i.e. we would need to know what block hash is the old/new one. Also, we might anyway need to check every log again before we dequeue it.

The block subscriber can tell what is the latest block history, which can be used by the buffer to remove logs from blocks that were reorged, regardless of sorting.